### PR TITLE
68 disolve sofastservice class

### PR DESCRIPTION
**Pull request by bbean23**
I'm so sorry, this became a very involved effort. But I added unit tests!

Anyways, I need apologize again because this is going to be gross to review. Sorry.

Key commits:
- new AbstractPlotHandler class
- move everything related to calibration, move projector image creation into ImageProjection, move camera exposure calibration into SofastService
- moving common code into functions, removing code that is less common
- use singleton design pattern for camera and projector

### CODE REVIEW

File: opencsp/app/scene_reconstruction/test/test_SceneReconstruction.py
Comment (braden6521) [inline opencsp/app/scene_reconstruction/test/test_SceneReconstruction.py:54]: Thanks for adding these lines.

File: opencsp/app/scene_reconstruction/test/test_SceneReconstruction.py
Comment (braden6521) [inline opencsp/app/scene_reconstruction/test/test_SceneReconstruction.py:57]: Is the output directory supposed to be "ImageAcquisitionAbstract" for this unit test? Could this be a copy/paste error?

Comment (bbean23) [reply opencsp/app/scene_reconstruction/test/test_SceneReconstruction.py:57]: Yeah, that's a copy paste error. Nice catch!

File: opencsp/common/lib/deflectometry/ImageProjection.py
Comment (braden6521) [inline opencsp/common/lib/deflectometry/ImageProjection.py:316]: Love this method, thanks for adding it. 
Just a couple of picky things: 
1. In the doc string, could we change "entire projected area" to "entire active area," or something similar to fit with the existing nomenclature better (see `display_image_in_active_area()`?
2. How do you feel about changing the method name to something more descriptive like `array_black_active_area()` or something like that?

Comment (bbean23) [reply opencsp/common/lib/deflectometry/ImageProjection.py:316]: 1. Yep, can do!
2. I chose "zeros" to match "np.zeros", but I can definitely change this to something else if you think that would be more appropriate. Maybe "get_black_array_active_area"?

Comment (braden6521) [reply opencsp/common/lib/deflectometry/ImageProjection.py:316]: Also it's totally my fault that the size x/y attributes are confusing; `size_win_x` is the size of the entire window, and `size_x` is just the size of the active area. If we wanted to clean that up, that might be appropriate, but don't feel obligated to do that in this PR.

File: opencsp/app/sofast/lib/SystemSofastFringe.py
Comment (braden6521) [inline opencsp/app/sofast/lib/SystemSofastFringe.py:445]: Would it make things simpler or more complicated if we made `fringes` an attribute of `SystemSofastFringe`? I'm thinking this could be beneficial because declaring the fringe object typically happens one/few times and the call to run the measurement sequence happens often. Might be marginal gains, let me know what you think.

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (braden6521) [inline opencsp/app/sofast/lib/sofast_common_functions.py:1]: This library is great, thanks for thinking of it! ❤️ I have a couple ideas of other things that might best belong here, see other comments. 
Only one picking thing: since it mostly relates to image acquisition and projection, what do you think about renaming this to something like: `sofast_comon_io_functions` or `sofast_common_system_functions` etc. for better clarity? Probably not necessary, and totally up to you!

File: opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py
Comment (braden6521) [inline opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py:1]: Should this entire file (test_ImageAcquisitionAbstract.py) be moved to common/lib/camera/test to live next to ImageAcquisitionAbstract.py?
(To be clear, I still think ImageAcquisition_no_camera.py should live here because it's used to test Sofast stuff fyi).

File: opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py
Comment (braden6521) [inline opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py:71]: Love this stub. Yes, we should definitely add this in the future.

File: opencsp/app/sofast/test/test_SystemSofastFringe.py
Comment (braden6521) [inline opencsp/app/sofast/test/test_SystemSofastFringe.py:17]: Does this really not need to be marked no_xvfb?

File: opencsp/common/lib/camera/ImageAcquisitionAbstract.py
Comment (braden6521) [inline opencsp/common/lib/camera/ImageAcquisitionAbstract.py:27]: I'm not really sure what's best, just throwing this out as a potential idea. But perhaps the cam_options() method belongs in sofast_common_functions()? I'm thinking this because I believe the only purpose of this dictionary of ImageAcquisition options is for the dropdown menu in SofastGUI (and perhaps other future GUIs too) to have the correct description and associated class. Lmk what you think.

File: opencsp/common/lib/camera/ImageAcquisitionAbstract.py
Comment (braden6521) [inline opencsp/common/lib/camera/ImageAcquisitionAbstract.py:42]: So, just to make things more complicated, as of now, there can be multiple ImageAcquisition instances in a SofastSystemFringe, (sometimes we use redundant cameras in a Sofast setup). You are definitely more knowledgeable about singleton design architecture, so I expect you already accounted for this or there is an easy patch. Perhaps something like `self._instance` -> `self._instances` where `_instances` is a list of instances?? 

Note that in ImageAcquisition_DCAM_xxx.py classes, the constructor takes in an "instance" integer that selects which of the Basler cameras to connect to.
https://github.com/sandialabs/OpenCSP/blob/c3128f0789aac55a02de39d337a7a173744eb507/opencsp/common/lib/camera/ImageAcquisition_DCAM_color.py#L10

It's probably a good idea to keep this functionality... but probably not 100% necessary. We can discuss what you think the best idea is.

File: opencsp/common/lib/deflectometry/ImageProjection.py
Comment (braden6521) [reply opencsp/common/lib/deflectometry/ImageProjection.py:316]: Yeah, matching `np.zeros` is totally legit. I like having "active_area" somewhere in the name though (I can be convinced either way though). I do like your suggestion above, `array_black_active_area()`!

File: opencsp/common/lib/deflectometry/ImageProjectionSetupGUI.py
Comment (braden6521) [inline opencsp/common/lib/deflectometry/ImageProjectionSetupGUI.py:379]: I like these changes!
YFI it is still on my list to change all `save_as_hdf` calls to `to_hdf`... just haven't gotten around to it yet.

File: opencsp/common/lib/render/figure_management.py
Comment (braden6521) [inline opencsp/common/lib/render/figure_management.py:369]: 🎉
Noted. I should probably start heeding this note.

File: opencsp/common/lib/target/TargetColor.py
Comment (braden6521) [inline opencsp/common/lib/target/TargetColor.py:7]: 🎉
Woo! Thanks for removing all these unused imports!! ❤️

File: opencsp/app/sofast/test/data/input/CameraCalibration/cal_global.h5
Comment (braden6521) [inline opencsp/app/sofast/test/data/input/CameraCalibration/cal_global.h5:1]: Not a big deal, but should these be stored in data/input/ImageCalibration (instead of CameraCalibration)? Could be confusing because there is a CameraCalibration app and this technically is for the ImageCalibrationXXX classes.

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (braden6521) [inline opencsp/app/sofast/lib/sofast_common_functions.py:1]: This library is great, thanks for thinking of it! ❤️ I have a couple ideas of other things that might best belong here, see other comments.
Only one picking thing: since it mostly relates to image acquisition and projection, what do you think about renaming this to something like: sofast_comon_io_functions or sofast_common_system_functions etc. for better clarity? Probably not necessary, and totally up to you!

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (braden6521) [inline opencsp/app/sofast/lib/sofast_common_functions.py:14]: I believe these are unused imports

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (braden6521) [inline opencsp/app/sofast/lib/sofast_common_functions.py:230]: I believe this should be `image_acquisition`, not `image_acquisition_default`

File: opencsp/app/sofast/lib/ImageCalibrationAbstract.py
Comment (bbean23) [reply opencsp/app/sofast/lib/ImageCalibrationAbstract.py:259]: I agree, these could be merged into a single class. They could also be left as is. Let's talk about it more tomorrow.

File: opencsp/common/lib/deflectometry/ImageProjection.py
Comment (bbean23) [reply opencsp/common/lib/deflectometry/ImageProjection.py:316]: Changed to `get_array_black_active_area` to match the "action" style of method naming conventions.

I agree that renaming `size_x`/`size_y` should happen at some point, just not in this PR.

File: opencsp/app/sofast/SofastGUI.py
Comment (braden6521) [inline opencsp/app/sofast/SofastGUI.py:72]: Looks like you edited SystemSofastFringe to only take "projector" as an input, not both projector and camera.

File: opencsp/app/sofast/lib/SystemSofastFringe.py
Comment (bbean23) [reply opencsp/app/sofast/lib/SystemSofastFringe.py:445]: I agree with your assertion that 'fringes' should be a property of SystemSofastFringe.  In fact, it already is!  I'm updating fringes to be get-only, and changing the method name to 'set_fringes'.

File: opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py
Comment (bbean23) [reply opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py:17]: They aren't used in this test.  I've been including them as essentially boiler-plate code, to make it easier to add data for tests later as needed.

So, do they need to be removed? Idk... 🤷‍♂️.  I'm in favor of keeping them, but I'm happy to remove them if you think that's better.

File: opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py
Comment (bbean23) [reply opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py:1]: Yes. Moved!

File: opencsp/app/sofast/test/test_SystemSofastFringe.py
Comment (bbean23) [reply opencsp/app/sofast/test/test_SystemSofastFringe.py:17]: I moved this decorator up to apply to whole class. Is that not correct?

File: opencsp/common/lib/camera/ImageAcquisitionAbstract.py
Comment (bbean23) [reply opencsp/common/lib/camera/ImageAcquisitionAbstract.py:42]: Yeah, good call. Let's keep the multiple-camera functionality.

This doesn't really fit in the singleton design pattern any more. What we can do instead is use the multiton design pattern:

```python
class ImageAcquisitionAbstract:
    _instances: dict[int,ImageAcquisitionAbstract] = {}
    """ All instantiated camera instances. We use a dictionary to ensure that once an index is assigned,
    then all references to that same index will return the same camera (or None if the camera was closed). """
    _next_instance_idx: int = 0
    """ The index to next use when adding a camera to the _instances dict. """

    @staticmethod
    def get_instance(idx: int=0) -> ImageAcquisitionAbstract:
        if idx in ImageAcquisitionAbstract._instances:
            camera = ImageAcquisitionAbstract[idx]
            if camera.is_closed:
                return None
        return None

    @staticmethod
    def get_instances(subclass: type[ImageAcquisitionAbstract]=None) -> list[ImageAcquisitionAbstract]:
        ret: list[ImageAcquisitionAbstract] = []
        for idx in self._instances:
            camera = ImageAcquisitionAbstract.get_instance(idx)
            if camera is None:
                continue
            if subclass is not None and not isinstance(camera, subclass):
                continue
            ret.append(camera)
        return ret
```

File: opencsp/common/lib/render/figure_management.py
Comment (bbean23) [reply opencsp/common/lib/render/figure_management.py:369]: Haha yes please. That will help make unit tests more reliable. :)

File: opencsp/app/sofast/test/data/input/CameraCalibration/cal_global.h5
Comment (bbean23) [reply opencsp/app/sofast/test/data/input/CameraCalibration/cal_global.h5:1]: Yes, I should have used that subdirectory instead. Moved!

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (bbean23) [reply opencsp/app/sofast/lib/sofast_common_functions.py:1]: I like that name more, but I'm not convinced yet that there won't be more types of "common functions" added later.  If you're ok with it, let's keep it as is and change it if it becomes more clear later that it should be changed.

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (bbean23) [reply opencsp/app/sofast/lib/sofast_common_functions.py:14]: They are indeed. Removed!

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (bbean23) [reply opencsp/app/sofast/lib/sofast_common_functions.py:230]: You had me worried, but `image_acquisition_default` is correct.

File: opencsp/app/sofast/lib/SystemSofastFringe.py
Comment (bbean23) [reply opencsp/app/sofast/lib/SystemSofastFringe.py:116]: Haha good catch. I did not. It should just be a `pass`.

File: opencsp/app/sofast/SofastGUI.py
Comment (bbean23) [reply opencsp/app/sofast/SofastGUI.py:72]: So I did. Fixed!

**Comment by bbean23** [comment]
Hey @braden6521, I think I've responded to all your comments. You were incredibly thorough and insightful, as always! Would you mind double-checking that I didn't miss anything, and approving if so? Thanks!

File: opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py
Comment (braden6521) [reply opencsp/app/sofast/test/test_ImageAcquisitionAbstract.py]: Yeah, that sounds good to me if it's a boiler plate section. I'm fine keeping them here.

File: opencsp/common/lib/camera/ImageAcquisitionAbstract.py
Comment (braden6521) [reply opencsp/common/lib/camera/ImageAcquisitionAbstract.py:32]: Sounds good.

File: opencsp/common/lib/camera/ImageAcquisitionAbstract.py
Comment (braden6521) [reply opencsp/common/lib/camera/ImageAcquisitionAbstract.py]: Sweet, nice work! Yes, multidon sounds like just what we want in this case.

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (braden6521) [reply opencsp/app/sofast/lib/sofast_common_functions.py:1]: Yeah, that sounds good. This library is hot of the presses, and I agree, likely to change in the near future.

File: opencsp/app/sofast/lib/sofast_common_functions.py
Comment (braden6521) [reply opencsp/app/sofast/lib/sofast_common_functions.py:227]: Are you sure (I could definitely be looking at the wrong function)? I'm seeing this inuat as `image_acquisition_default`

https://github.com/bbean23/OpenCSP/blob/ca7a69b915eb6baa522645d0d09f742ecc84d1da/opencsp/app/sofast/lib/sofast_common_functions.py#L73

**Comment by braden6521** [comment]
@bbean, looks good. Could you double check this comment one more time.
https://github.com/sandialabs/OpenCSP/pull/72#discussion_r1557962995

And if you don't mind, maybe we could give this a test in the optics lab before we merge.

**Review by braden6521 — APPROVED** [review]
Approved!

**Pull request merged**
